server/routerlicious/packages/services/src/redis.ts

```diff
@@ -81,10 +100,27 @@ export class RedisCache implements ICache {
 		}
 	}
 
+	/**
+	 * Get a list of keys that have a given prefix.
+	 *
+	 * @param keyPrefix - Prefix for the keys to get.
+	 */
+	public async keysByPrefix(keyPrefix: string): Promise<string[]> {
+		return this.client.keys(`${this.getKey(keyPrefix)}*`);
+	}
+
 	/**
 	 * Translates the input key to the one we will actually store in redis
+	 *
+	 * @param key - The input key
+	 * @param prefix - Prefix to append to key. Empty string will not add any prefix to the key.
 	 */
-	private getKey(key: string): string {
-		return `${this.prefix}:${key}`;
+	private getKey(key: string, prefix?: string): string {
```

Comment:
I would suggest to not inject two prefixes, one from constructor one from method, it would be hard to understand.
Just to clarify:
1. Keep the one in constructor would be surfficient.
2. If you REALLY want to keep both, then rename the prefix in method as `prefixOverride`

Reply:
The reason I allow for a prefix override is because of certain use-cases where you have a class that takes an ICache as a constructor argument, which means you would not know what the prefix of the passed in cache is. 

A specific case where this will happen is in the DocumentManager changes I am making, where document manager caches document static properties in redis. Document manager is in server-services, and is meant to be used across r11s, historian, and gitrest. However the RedisCaches for each of these will have each been initialized with a different prefix from their configs (like "page" for r11s). By allowing prefix overrides, a class like DocumentManager could just override the prefix for all of its calls, to ensure that only one cache is being stored. This also prevents a cache miss when a DocumentManager with different cache prefix reads static data for the first time.

I agree that the name is confusing though, will change to prefixOverride.

Reply:
I still think this is a little bit confusing. You can reference on this page: https://nickcraver.com/blog/2019/08/06/stack-overflow-how-we-do-app-caching/

Maybe we could do:
1. component prefix, that is the config you set up in `redisConfig` and we always prefix all keys with component prefix
2. key prefix, that used in `setKey` `getKey`, that a caller can use if they want to.

Basically, a key will be format like `${compoenentPrefix}_${keyPrefix}_{key}`, we will get ride of these override/existance/nonexistance manament thing. If all prefix not exist, a key will just be `__key`, I think it still looks OK to me. Thought?

Reply:
I think that the issue with the component prefix is the same as before though. If I understand correctly, a RedisCache created in Historian and R11s would have have different componentPrefixes. If you tried using these two different RedisCache instances to store/retrieve the same data, the two keys would look something like: "historian_keyPrefix_key" and "r11s_keyPrefix_key". This would lead to them storing/accessing different data, which is not always necessary. For the Document static data caching, it would be better if both of them could just use one key. Since DocumentManager is in r11s, the key prefix overrides could let you override any componentPrefix, so that any key the DocumentManager caches looks like "r11s_keyPrefix_key" (even if it was passed a Historian RedisCache).

But tomorrow I'll look into this some more, and also read through the rest of the blog post that you linked

Reply:
Also, if this helps to explain it, this PR also makes some changes to documentManager.ts.
In these changes, DocumentManager takes an ICache object into its constructor. 
DocumentManager in these changes also has a static variable, cacheKeyPrefix, that overrides the key prefix for all of the RedisCache calls made by the class. This is so that no matter which prefix the RedisCache passed into DocumentManager uses, DocumentManager will always store a single key for each documentId, instead of multiple data entries per documentId for each possible RedisCache prefix.

Reply:
Let's have an offline discussion to be more productive. :)

Reply:
Resolved as per offline discussion